apps/designer/app/designer/features/style-panel/controls/text/text-control.tsx

```diff
@@ -0,0 +1,228 @@
+    currentStyle,
+    inheritedStyle,
+    property: styleConfig.property,
+  });
+
+  if (value === undefined) return null;
+
+  // @todo setValue has a latency issue when updating a value fast(i.e keyboard arrows/scrub control)
+  const setValue = setProperty(styleConfig.property);
+
+  const Icon = iconConfigs[styleConfig.property]?.normal;
+
+  return (
+    <Grid
+      css={{
+        gridTemplateColumns: "$sizes$6 auto $sizes$6",
+        gridTemplateRows: "repeat(1, 1fr)",
+      }}
+    >
+      <Tooltip
+        content={styleConfig.label}
+        delayDuration={200}
+        disableHoverableContent={true}
+      >
+        <IconButton
+          onPointerEnter={({ target }) => {
+            if (value.type !== "unit") return;
+            numericScrubRef.current =
+              numericScrubRef.current ||
+              numericScrubControl(target as HTMLInputElement, {
+                initialValue: value.value as number,
+                onValueChange: ({ value }) => setValue(String(value)),
+              });
+          }}
+          onPointerLeave={() => {
+            if (value.type !== "unit") return;
+            numericScrubRef.current?.disconnectedCallback();
+            numericScrubRef.current = undefined;
+          }}
+          variant="ghost"
+          size="1"
+          css={{
+            ...CSSGridLeft,
+            ...CSSIconButton,
+            ...(isCurrentBreakpoint && {
+              bc: "$colors$blue4",
+              "&:not(:hover)": {
+                color: "$colors$blue11",
+              },
+            }),
+          }}
+        >
+          {Icon && <Icon />}
+        </IconButton>
+      </Tooltip>
+      <TextField
+        css={CSSTextField}
+        spellCheck={false}
+        value={value.value}
+        onChange={(event: ChangeEvent<HTMLInputElement>) =>
+          setValue(event.target.value)
+        }
+        onFocus={(event: FocusEvent<HTMLInputElement>) => event.target.select()}
+        onKeyDown={(event: KeyboardEvent<HTMLInputElement>) => {
+          if (value.type !== "unit") return;
+          if (!["ArrowUp", "ArrowDown"].includes(event.code)) return;
+          event.preventDefault();
+          let currentValue = value.value;
+          let currentDelta = 1;
+          if (event.shiftKey) currentDelta = 10;
+          if (event.altKey) currentDelta = 0.1;
+          if (event.code === "ArrowUp")
+            currentValue = currentValue + currentDelta;
+          if (event.code === "ArrowDown")
+            currentValue = currentValue - currentDelta;
+          const currentValueAsString =
+            currentValue % 1
+              ? currentValue.toPrecision(
+                  Math.abs(currentValue).toString().indexOf(".") + 2
+                )
+              : String(currentValue);
+          setValue(currentValueAsString);
+        }}
+      />
+      {value.type === "unit" ? (
+        <Units
+          value={value.unit}
+          items={units
```

Comment:
we can extract that this entire thing where you map reduce and map the units into a static array above, so it happens just once

Reply:
and calling it something like menutUnitsSubset would explain why we have a shortlist